README
This is a "plugin" for the Video Disk Recorder (VDR).

Originally written by:                  Thomas Koch <tom@linvdr.org>
Development since 0.8:                  Michael Schneider <vdrportal_midas <at> gmx <dot> de>
Old project's homepage:          http://www.linvdr.org/projects/sleeptimer/
New homepage:                    http://projects.vdr-developer.org/projects/plg-sleeptimer

Latest version available at: Project's homepage

See the file COPYING for license information.

*******************************
*** Command Line Parameters ***
*******************************

use -e parameter to redefine the command for shutdown. By default /sbin/poweroff is called


********************
*** Plugin Setup ***
********************

Default Timespan [min]:
-defines the default timespan till shutdown, which is shown in the plugins main menu entry

Action:
-Shutdown represents the command to be executed on sleeptimer timeout (default: /sbin/poweroff)
 overridden by command line parameter -e (see above)
-Mute does not execute any command, but mutes audio if not already muted

Use Autoswitch with default:
-Integrates the default timespan in the predefined values and makes the Autoswitch function
 start with this value.

Confirm:
-makes the Autoswitch function offer the predefined timespans but does not activate
 the sleeptimer unless the user explicitly confirms by pressing Ok.


*****************
*** Main Menu ***
*****************

Not Active / Disable sleeptimer:
-Allows disabling of an active sleeptimer. If there is no active sleeptimer 'Not Active' is
 displayed.

Shutdown [min]:
-Denotes the timespan for a sleeptimer in minutes. Values < 2 minutes are not allowed. 
 Activation by Ok.

Shutdown (time):
-Denotes the time for a sleeptimer. Values < 2 minutes are not allowd. Activation by Ok.

Red / Autoswitch (inspired by an idea / feature request by grindo):
-Switches through a set of predefined values (min). Especially convenient in combination with
 the use of keymacros. Just add for example
 User7     @sleeptimer Red
 in vdr's keymacros.conf file and make sure User7 is mapped in remote.conf. This makes the
 user switch through the predefined values by just pressing one key and without hassling with
 vdr's main menu. By default the predefined values are:
 Disabling, 15, 30, 45, 60, 90, 120, 180
 In combination with 'Use default for Autoswitch' the 'Default Timespan' value will be
 integrated in this sequence and calling the autoswitch function will start with this
 value.

Green / Disable:
-Disables an active sleeptimer.


debian/changelog
vdr-plugin-sleeptimer (0.8.3-201205011650dev) experimental; urgency=low

  * Makefile adaption concerning i18n. Thx to idl0r.

 -- Midas <vdrportal_midas <at> gmx <dot> de>  Tue, 01 May 2012 16:51:46 +0200

vdr-plugin-sleeptimer (0.8.3-201205011141dev) experimental; urgency=low

  * added missing include to i18n.h. thanx to nox again
  * changed -Woverloaded-virtual to -Werror=overloaded-virtual in Makefile

 -- Midas <vdrportal_midas <at> gmx <dot> de>  Tue, 01 May 2012 11:41:30 +0200

vdr-plugin-sleeptimer (0.8.3-DEV) experimental; urgency=low

  * Non-maintainer upload.
  * Developer version (WIP). Don't expect it to work.

 -- Midas <vdrportal_midas <at> gmx <dot> de>  Sat, 06 Aug 2011 13:01:26 +0200

vdr-plugin-sleeptimer (0.8.2) experimental; urgency=low

  * Non-maintainer upload.
  * New upstream release.

 -- Midas <vdrportal_midas <at> gmx <dot> de>  Tue, 02 Aug 2011 02:48:53 +0200

vdr-plugin-sleeptimer (0.8.1) experimental; urgency=low

  * Non-maintainer upload.
  * New upstream release

 -- Midas <vdrportal_midas at gmx dot de>  Sun, 31 Jul 2011 09:08:16 +0200

vdr-plugin-sleeptimer (0.8-1) experimental; urgency=low

  * Non-maintainer upload.
  * New upstream release

 -- Midas <vdrportal_midas at gmx dot de>  Thu, 21 Jul 2011 07:29:16 +0200

vdr-plugin-sleeptimer (0.7-21) experimental; urgency=low

  * Removed non-standard shebang line from debian/rules
  * Added README.source
  * Standards-Version: 3.8.3

 -- Tobias Grimm <etobi@debian.org>  Mon, 09 Nov 2009 20:52:24 +0100

vdr-plugin-sleeptimer (0.7-20) experimental; urgency=low

  * Release for vdr 1.7.6
  * Added ${misc:Depends}
  * Bumped standards version to 3.8.1
  * Changed section to "video"

 -- Thomas Günther <tom@toms-cafe.de>  Wed, 29 Apr 2009 23:11:31 +0200

vdr-plugin-sleeptimer (0.7-19) experimental; urgency=low

  * Dropped patchlevel control field
  * Build-Depend on vdr-dev (>=1.6.0-5)
  * Bumped Standards-Version to 3.8.0
  * Switched Build-System to cdbs, Build-Depend on cdbs
  * Using COMPAT=5 now

 -- Tobias Grimm <tg@e-tobi.net>  Fri, 25 Jul 2008 15:54:28 +0200

vdr-plugin-sleeptimer (0.7-18) experimental; urgency=low

  * Increased package version to force rebuild for vdr 1.6.0-1ctvdr7

 -- Tobias Grimm <tg@e-tobi.net>  Mon, 12 May 2008 00:56:58 +0200

vdr-plugin-sleeptimer (0.7-17) experimental; urgency=low

  * Build-Depend on vdr-dev (>= 1.6.0)

 -- Tobias Grimm <tg@e-tobi.net>  Mon, 24 Mar 2008 20:14:27 +0100

vdr-plugin-sleeptimer (0.7-16) experimental; urgency=low

  * Force rebuild for vdr 1.5.15

 -- Tobias Grimm <tg@e-tobi.net>  Mon, 18 Feb 2008 21:33:29 +0100

vdr-plugin-sleeptimer (0.7-15) unstable; urgency=low

  * Release for vdr 1.5.13

 -- Thomas Günther <tom@toms-cafe.de>  Wed, 16 Jan 2008 10:31:50 +0100

vdr-plugin-sleeptimer (0.7-14) unstable; urgency=low

  * Release for vdr 1.5.12

 -- Thomas Günther <tom@toms-cafe.de>  Tue, 20 Nov 2007 23:46:48 +0100

vdr-plugin-sleeptimer (0.7-13) unstable; urgency=low

  * Release for vdr 1.5.11

 -- Thomas Günther <tom@toms-cafe.de>  Tue,  6 Nov 2007 23:34:38 +0100

vdr-plugin-sleeptimer (0.7-12) unstable; urgency=low

  * Release for vdr 1.5.10

 -- Thomas Günther <tom@toms-cafe.de>  Tue, 16 Oct 2007 23:51:17 +0200

vdr-plugin-sleeptimer (0.7-11) unstable; urgency=low

  * Release for vdr 1.5.9

 -- Thomas Günther <tom@toms-cafe.de>  Tue, 28 Aug 2007 01:01:28 +0200

vdr-plugin-sleeptimer (0.7-10) unstable; urgency=low

  * Release for vdr 1.5.8

 -- Thomas Günther <tom@toms-cafe.de>  Thu, 23 Aug 2007 01:09:31 +0200

vdr-plugin-sleeptimer (0.7-9) unstable; urgency=low

  * Release for vdr 1.5.6

 -- Thomas Günther <tom@toms-cafe.de>  Tue, 14 Aug 2007 01:46:40 +0200

vdr-plugin-sleeptimer (0.7-8) unstable; urgency=low

  * Release for vdr 1.5.5

 -- Thomas Günther <tom@toms-cafe.de>  Wed, 27 Jun 2007 23:03:10 +0200

vdr-plugin-sleeptimer (0.7-7) unstable; urgency=low

  * Release for vdr 1.5.2

 -- Thomas Günther <tom@toms-cafe.de>  Sat, 28 Apr 2007 00:01:19 +0200

vdr-plugin-sleeptimer (0.7-6) unstable; urgency=low

  * Release for vdr 1.5.1

 -- Thomas Günther <tom@toms-cafe.de>  Tue, 27 Feb 2007 19:59:38 +0100

vdr-plugin-sleeptimer (0.7-5) unstable; urgency=low

  [ Thomas Günther ]
  * Replaced VDRdevel adaptions in debian/rules with make-special-vdr
  * Adapted call of dependencies.sh and patchlevel.sh to the new location
    in /usr/share/vdr-dev/

  [ Tobias Grimm ]
  * Build-Depend on vdr-dev (>=1.4.5-1)

 -- Tobias Grimm <tg@e-tobi.net>  Sun, 14 Jan 2007 19:17:49 +0100

vdr-plugin-sleeptimer (0.7-4) unstable; urgency=low

  * Build-Depend on vdr-dev (>=1.4.4-1)

 -- Tobias Grimm <tg@e-tobi.net>  Sun,  5 Nov 2006 11:45:43 +0100

vdr-plugin-sleeptimer (0.7-3) unstable; urgency=low

  * Build-Depend on vdr-dev (>=1.4.3-1)

 -- Tobias Grimm <tg@e-tobi.net>  Sat, 23 Sep 2006 20:33:00 +0200

vdr-plugin-sleeptimer (0.7-2) unstable; urgency=low

  * Build-Depend on vdr-dev (>=1.4.2-1)
  * Bumped Standards-Version to 3.7.2

 -- Tobias Grimm <tg@e-tobi.net>  Sun, 27 Aug 2006 14:39:35 +0200

vdr-plugin-sleeptimer (0.7-1) unstable; urgency=low

  * Tobias Grimm <tg@e-tobi.net>
    - New upstream release
    - Removed 02_sleeptimer-1.3.32.dpatch
    - Removed 90_APIVERSION.dpatch

 -- Debian VDR Team <pkg-vdr-dvb-devel@lists.alioth.debian.org>  Sun, 18 Jun 2006 16:16:58 +0200

vdr-plugin-sleeptimer (0.6-15) unstable; urgency=low

  * Thomas Schmidt <tschmidt@debian.org>
    - Build-Depend on vdr-dev (>=1.4.1-1)

 -- Debian VDR Team <pkg-vdr-dvb-devel@lists.alioth.debian.org>  Fri, 16 Jun 2006 22:08:01 +0200

vdr-plugin-sleeptimer (0.6-14) unstable; urgency=low

  * Thomas Schmidt <tschmidt@debian.org>
    - Build-Depend on vdr-dev (>=1.4.0-1)

 -- Debian VDR Team <pkg-vdr-dvb-devel@lists.alioth.debian.org>  Wed,  3 May 2006 23:08:25 +0200

vdr-plugin-sleeptimer (0.6-13) unstable; urgency=low

  * Tobias Grimm <tg@e-tobi.net>
    - Build-Depend on vdr (>=1.3.48-1)
    - Added 90_APIVERSION.dpatch


sleeptimer.c
/*
 * sleeptimer.c: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id$
 */

#include <vdr/plugin.h>
#include <vdr/interface.h>
#include <vdr/device.h>
#include <vdr/status.h>

#if VDRVERSNUM < 10507
 #include "i18n.h"
#endif 

#ifndef COMMANDLENGTH
 #define COMMANDLENGTH 512
#endif

static const char *VERSION        = "0.8.3-201205011650dev";
static const char *DESCRIPTION    = "Sleeptimer for VDR";
static const char *MAINMENUENTRY  = tr("Sleeptimer");

int method = 0;
int default_timespan=15;
int shutdown_time = 0;
int shutdown_minutes = 2;
bool start_w_default=false;
bool do_confirm=false;
char command[COMMANDLENGTH] = "/sbin/poweroff";
char cli_command[COMMANDLENGTH]="";
bool process_red=false;

#define ARRAYLENGTH 9
int autoswitch_vals[] = { 0,15,30,45,60,90,120,180,-1 }; //-1 is a placeholder for the default timespan


// Tools
int i2s(int val) //prepare a special int for vdrs OSD
{
 time_t tme = time(NULL);
 struct tm *TM;
 int dif;
 int h;
 int min;

 TM = localtime(&tme);
 (*TM).tm_sec = 0;
 (*TM).tm_min = 0;
 (*TM).tm_hour = 0;
 dif = (val - mktime(TM));
 h = dif / 3600;
 min = ((dif - h * 3600) / 60);
 if (h > 23) h-=24;
 return h * 100 + min; //Std * 100 because vdr interprets the value in a special way (eg 10:46h in vdr is 1046)
}

time_t s2i(int &min, int &h, int value) //convert vdrs representation of time (see i2s) to time_t and check
{                                         //if the shutdown time is after midnight
 time_t tme = time(NULL);
 struct tm *TM;
 TM = localtime(&tme);
 h = int(value / 100);
 min = ((value - h * 100) % 60);

 if (h * 60 + min < (*TM).tm_hour * 60 + (*TM).tm_min)
  (*TM).tm_mday++;

 (*TM).tm_sec = 0;
 (*TM).tm_min = 0;
 (*TM).tm_hour = 0;
 return mktime(TM) + h * 3600 + min * 60;
}

void InsertDefaultTimespan()
{
 int reinit_array[] = { 0, 15, 30, 45, 60, 90, 120, 180, -1};
 int i=0;
 bool already_in=false;
 for (i=0;i<ARRAYLENGTH;i++) 
 {
  autoswitch_vals[i]=reinit_array[i];
  if (default_timespan==reinit_array[i]) already_in=true;
 }
 if (already_in) return;
 
 if (default_timespan>autoswitch_vals[ARRAYLENGTH-2])
 {
  autoswitch_vals[ARRAYLENGTH-1]=default_timespan;
  return;
 }
 else
 {
  for (i=ARRAYLENGTH-2;autoswitch_vals[i]>default_timespan;i--);
  int j=ARRAYLENGTH-1;
  for (;j>i+1;j--) autoswitch_vals[j]=autoswitch_vals[j-1];
  autoswitch_vals[i+1]=default_timespan;
 }
}

void ResetAutoswitchValues()
{
 int reinit_array[] = { 0, 15, 30, 45, 60, 90, 120, 180, -1};
 int i=0;
 for (i=0;i<ARRAYLENGTH;i++) 
 {
  autoswitch_vals[i]=reinit_array[i];
 }
}
 
// cPluginSleeptimer
class cPluginSleeptimer : public cPlugin {


private:
  static time_t sleepat;
  time_t lastaction;
  void HotkeyAction();
  
public:
  cPluginSleeptimer(void);
  virtual ~cPluginSleeptimer();
  virtual const char *Version(void) { return VERSION; }
  virtual const char *Description(void) { return DESCRIPTION; }
  virtual const char *CommandLineHelp(void);
  virtual bool ProcessArgs(int argc, char *argv[]);
  virtual bool Start(void);
  virtual void Housekeeping(void);
  virtual const char *MainMenuEntry(void) { return MAINMENUENTRY; }
  virtual cOsdObject *MainMenuAction(void);
  virtual cMenuSetupPage *SetupMenu(void);
  virtual bool SetupParse(const char *Name, const char *Value);

  static time_t getSleeptimer() { return cPluginSleeptimer::sleepat; }
  static void setSleeptimer(time_t tt) { cPluginSleeptimer::sleepat=tt; }
  static eKeys ShowMessage(cString msg);
};

cPluginSleeptimer::cPluginSleeptimer(void)
{
  // Initialize any member variables here.
  // DON'T DO ANYTHING ELSE THAT MAY HAVE SIDE EFFECTS, REQUIRE GLOBAL
  // VDR OBJECTS TO EXIST OR PRODUCE ANY OUTPUT!
  lastaction=0;
}

//Initialization of static members
time_t cPluginSleeptimer::sleepat=0;


cPluginSleeptimer::~cPluginSleeptimer()
{
  // Clean up after yourself!
}

const char *cPluginSleeptimer::CommandLineHelp(void)
{
  // Return a string that describes all known command line options.
	return "  -e command   shutdown command (default: /sbin/poweroff)\n";
}

bool cPluginSleeptimer::ProcessArgs(int argc, char *argv[])
{
  // Implement command line argument processing here if applicable.
	int c;
	if (!strcmp(*argv, "sleeptimer"))
		++argv, --argc;
	while((c = getopt(argc, argv, "e:")) != -1 ) {
		switch(c) {
			case 'e': 
				strncpy(cli_command, optarg, sizeof(cli_command) - 1);
				break;
			default: return false;
		}
	}
  return true;
}

bool cPluginSleeptimer::Start(void)
{
#if VDRVERSNUM < 10507 
 RegisterI18n(Phrases);
#endif
 return true;
}

void cPluginSleeptimer::Housekeeping(void)
{
}

// cBackgroundSleeptimer 
class cBackgroundSleeptimer : public cThread {
private:
	virtual void Action(void);
	static cBackgroundSleeptimer *inst;
	cBackgroundSleeptimer(void);
	
public:
	static cBackgroundSleeptimer *getInstance(void);
	~cBackgroundSleeptimer();
};

class MainMenu : public cOsdMenu {
private: 
        int minpos;
        int timepos;
        int actionpos;

public:
	MainMenu(void);
	eOSState ProcessKey(eKeys k);
        ~MainMenu();
};


MainMenu::MainMenu(void) : cOsdMenu("Sleeptimer", 20) 
{
 if (process_red) return;
 char buf[80];
 SetHelp(tr("Autoswitch"),tr("Disable"),NULL,NULL);
 if(!cPluginSleeptimer::getSleeptimer()) 
 {
  snprintf(buf, sizeof(buf), "%s", tr("Not active"));
  shutdown_time = i2s((int)(time(NULL))+120);
  shutdown_minutes = default_timespan;
  Add(new cMenuEditIntItem(tr("Shutdown [min]"), &shutdown_minutes, 2,1440));
  Add(new cMenuEditTimeItem(tr("Shutdown (time)"), &shutdown_time));
  Add(new cOsdItem(hk(buf)));
  minpos=0;
  timepos=1;
  actionpos=2;
 }
 else
 {
  int dif=(int)(difftime(cPluginSleeptimer::getSleeptimer(),time(NULL)));

  shutdown_minutes = dif/60;

  //make sure shutdown_time (in the OSD) is always at least 2 minutes in the future
  int shutd_tmp=0;
  if (shutdown_minutes==0) shutd_tmp=2;
  if (shutdown_minutes==1) shutd_tmp=1;
  shutdown_time = i2s((int)cPluginSleeptimer::getSleeptimer()+shutd_tmp);

  snprintf(buf, sizeof(buf), tr("Disable sleeptimer in %d minutes"), shutdown_minutes);
  Add(new cOsdItem(hk(buf)));
  Add(new cMenuEditIntItem(tr("Shutdown [min]"), &shutdown_minutes,2,1440));
  Add(new cMenuEditTimeItem(tr("Shutdown (time)"), &shutdown_time));
  minpos=1;
  timepos=2;
  actionpos=0;
 }
}


cOsdObject *cPluginSleeptimer::MainMenuAction(void)
{
 if (Interface->GetKey(false)==kNone && !process_red) return new MainMenu();
 
 if (process_red) process_red=false;
 
 HotkeyAction();
 
 return NULL;

}

class cMenuSetupSleeptimer : public cMenuSetupPage {
private:
	int new_method;
	int timespan;
	int setup_start_w_default;
	int setup_do_confirm;
	char setup_command[COMMANDLENGTH];
	const char* methods[3];
public:
	cMenuSetupSleeptimer(void);
	virtual void Store(void);
};

cMenuSetupPage *cPluginSleeptimer::SetupMenu(void)
{
  return new cMenuSetupSleeptimer;
}

bool cPluginSleeptimer::SetupParse(const char *Name, const char *Value)
{
 if(!strcasecmp(Name, "Method"))
  method = atoi(Value);
 else if (!strcasecmp(Name,"DefaultTimespan"))
      {
       default_timespan=atoi(Value);
       if (start_w_default)
        InsertDefaultTimespan();
       else 
        ResetAutoswitchValues();
      }
      else if (!strcasecmp(Name,"StartWithDefault"))
           {
            start_w_default=atoi(Value);
            if (start_w_default)                 
             InsertDefaultTimespan();
            else 
             ResetAutoswitchValues();
           }
           else if (!strcasecmp(Name,"Confirmation"))
            do_confirm=atoi(Value);
           else if (!strcasecmp(Name,"Command"))
	strncpy(command, Value, sizeof(command) - 1);
 else return false;
 return true;
}

 
// cMenuSetupSleeptimer
cMenuSetupSleeptimer::cMenuSetupSleeptimer(void) 
{
 new_method = method;
 timespan = default_timespan;
 setup_start_w_default=start_w_default;
 strncpy(setup_command, command, sizeof(setup_command) - 1);
 setup_do_confirm=do_confirm;
 
 methods[0]=tr("Shutdown");
 methods[1]=tr("Mute");
 methods[2]=tr("Execute command");
 
 char* allowed_chars=(char*)" abcdefghijklmnopqrstuvwxyz0123456789-.,#~\\^$[]|()*+?{}/:%@&";
 
 Add(new cMenuEditIntItem(tr("Default Timespan [min]"),  &timespan, 2));
 Add(new cMenuEditStraItem(tr("Action"), &new_method, 3, methods)); 
 
 cMenuEditStrItem *commanditem;
 if (cli_command[0]=='\0')
 {
  commanditem = new cMenuEditStrItem(tr("Command"), setup_command, COMMANDLENGTH, allowed_chars);
 }
 else
 {
  commanditem = new cMenuEditStrItem(tr("Command"), cli_command, COMMANDLENGTH, allowed_chars);
  commanditem->SetSelectable(false);
 }
 Add(commanditem);
  
 Add(new cMenuEditBoolItem(tr("Use default for autoswitch"), &setup_start_w_default, tr("No"), tr("Yes")));
 Add(new cMenuEditBoolItem(tr("Confirm"), &setup_do_confirm, tr("No"), tr("Yes")));
}

void cMenuSetupSleeptimer::Store(void) 
{
 SetupStore("Method", method = new_method);
 SetupStore("DefaultTimespan", default_timespan=timespan);
 strncpy(command, setup_command, sizeof(command) - 1);
 SetupStore("Command", command);
 SetupStore("StartWithDefault", start_w_default=setup_start_w_default);
 SetupStore("Confirmation", do_confirm=setup_do_confirm);
 if (start_w_default) 
  InsertDefaultTimespan();
 else 
  ResetAutoswitchValues();
}


MainMenu::~MainMenu()
{
}

cBackgroundSleeptimer::cBackgroundSleeptimer(void) 
{
 Start();
}

cBackgroundSleeptimer* cBackgroundSleeptimer::inst=NULL;

cBackgroundSleeptimer* cBackgroundSleeptimer::getInstance()
{
 if (inst==NULL)
 {
  inst=new cBackgroundSleeptimer();
 }
 return inst;
}
 

cBackgroundSleeptimer::~cBackgroundSleeptimer() 
{
 if (inst) delete inst;
 cPluginSleeptimer::setSleeptimer(0);
}

void cBackgroundSleeptimer::Action(void) 
{
 isyslog("plugin-sleeptimer: thread started (pid=%d)", getpid());
 while(cPluginSleeptimer::getSleeptimer())
 {
  if(cPluginSleeptimer::getSleeptimer() <= time(NULL)) 
  {
   isyslog("plugin-sleeptimer: timeout");
   cPluginSleeptimer::setSleeptimer(0);

   if(method == 0) 
   {
    isyslog("plugin-sleeptimer: pressing key Power internally");
    if(!(cRemote::Put(kPower)))
     isyslog("plugin-sleeptimer: ERROR: internal keypress did not succeed \"%s\"!", command);
   }

   if(method == 1) 
   {
    isyslog("plugin-sleeptimer: muting audio");
    if(!cDevice::PrimaryDevice()->IsMute())
     cDevice::PrimaryDevice()->ToggleMute();
   }

   if(method == 2) 
   {
    dsyslog("plugin-sleeptimer: executing \"%s\"", command);
    if(SystemExec(command) == -1)
     isyslog("plugin-sleeptimer: ERROR while executing \"%s\"!", command);
   }

  } 
  else //timeout not reached yet
  {
   if((cPluginSleeptimer::getSleeptimer() - 60) <= time(NULL)) 
   {
    Skins.QueueMessage(mtInfo,tr("Going to sleep in about one minute"),0,-1);
    isyslog("plugin-sleeptimer: going to sleep in about one minute");
   }
  }
  int i=0;
  while (cPluginSleeptimer::getSleeptimer() && i<10)
  {
   sleep(1);
   i++;
  }
 }
 isyslog("plugin-sleeptimer: thread end (pid=%d)", getpid());
}


eOSState MainMenu::ProcessKey(eKeys k) 
{
 int current;

 eOSState state = cOsdMenu::ProcessKey(k); 
 switch(state) 
 {
  case osUnknown:
   switch(k) 
   {
    case kRed:
     process_red=true;
     cRemote::CallPlugin("sleeptimer");
     return osEnd;
     break;
    
    case kGreen:
     cPluginSleeptimer::setSleeptimer(0);
     return osEnd;
     break;

    case kOk:
     current = Current();
     if(current == actionpos) 
     {
      if(Interface->Confirm(tr("Disable sleeptimer?"))) 
      {
       cPluginSleeptimer::setSleeptimer(0);
       return(osEnd);
      }
     } 
     else if(current == timepos) 
     {
      char buf[80];
      int Std, Min;
      time_t tmp;
      tmp = s2i(Min, Std, shutdown_time);
                       
      snprintf(buf, sizeof(buf), tr("Activate sleeptimer at %i:%.2i?"), Std, Min);
           
      if(Interface->Confirm(buf)) 
      {
       if (tmp<time(NULL)+61)
       {
        Skins.Message(mtError,tr("Timespan shorter than 2 minutes")); return
        osContinue;
       }
             
       cPluginSleeptimer::setSleeptimer(tmp);

       cBackgroundSleeptimer::getInstance()->Start();
       return(osEnd);
      }
     } 
     else if(current == minpos) 
     {
      char buf[80];
      snprintf(buf, sizeof(buf), tr("Activate sleeptimer in %i minutes?"), shutdown_minutes);
      if(Interface->Confirm(buf)) 
      {
       cPluginSleeptimer::setSleeptimer(time(NULL) + shutdown_minutes * 60);
       cBackgroundSleeptimer::getInstance()->Start();
       return(osEnd);
      }
     }
     return(osPlugin);
     break;
     
    default: break;
   }
   default: break;
  }
  return state;
}


//   ************************
//   ***** HotkeyAction *****
//   ************************

void cPluginSleeptimer::HotkeyAction()
{
 eKeys user_key=(eKeys)NULL;
 cString msg="";
 bool first_action=false;
 time_t now=time(NULL);
 int next_sleepmin_index=-1;

 while (true) //OSD message / user interaction loop
 {
  now=time(NULL);
  if (difftime(now,lastaction)>5) first_action=true;

  lastaction=now;

  if (first_action)
  {
   next_sleepmin_index=-1;
   if (sleepat)
   {
    int sleepint=(int)(difftime(sleepat,now)/60);
    if (sleepint==1)
    {
     msg=msg.sprintf(tr("Sleeptimer in %i minute"),sleepint);
    }
    else
    {
     msg=msg.sprintf(tr("Sleeptimer in %i minutes"),sleepint);
    }
    
   }
   else
   {
    msg=msg.sprintf(tr("Sleeptimer disabled"));
   }
  }
  else
  {
   if (next_sleepmin_index > 0)
   {
    if (do_confirm)
    {
     msg=msg.sprintf(tr("Enable sleeptimer in %i minutes?"), autoswitch_vals[next_sleepmin_index]);
    }
    else
    {
     msg=msg.sprintf(tr("Sleeptimer in %i minutes"), autoswitch_vals[next_sleepmin_index]);
    }
   } 
   else
   {
    if (do_confirm)
    {
     msg=msg.sprintf(tr("Disable sleeptimer?"));
    }
    else
    {
     msg=msg.sprintf(tr("Sleeptimer disabled"));
    }
   } 
  }
  eKeys k=cPluginSleeptimer::ShowMessage(msg);


// ***********************
// ***** KEYHANDLING *****
// ***********************

  if (first_action)
  {
   if ((k>=kUser1 && k<=kUser9) || k==kRed)
    user_key=k;
  }

  if (k!=kRed && k!=kGreen && k!=kBack && k!=kOk && k!= kNone && k!=user_key) 
  {
   cRemote::Put(k);
   break;
  }

  if (k==kBack) 
  {
   break;
  }
 
  if (k==kRed || k==user_key)
  {
   if (!first_action)
   {
    if (!do_confirm)
    {
     if (next_sleepmin_index > 0)
     {
      sleepat=now + autoswitch_vals[next_sleepmin_index] * 60;
      cBackgroundSleeptimer::getInstance()->Start();
     }
     else
     {
      sleepat=0;
     }
    }
   }
  }


  if (k==kOk || (k==kNone && !do_confirm))
  {
   if (first_action) break;
   if (next_sleepmin_index > 0)
   {
    sleepat=now + autoswitch_vals[next_sleepmin_index] * 60;
    cBackgroundSleeptimer::getInstance()->Start();
   }
   else
   {
    sleepat=0;
   }
   break;
  }

  if (k==kGreen)
  {
   sleepat=0;
   msg=msg.sprintf(tr("Sleeptimer disabled"));
   break;
  }
  
  if (k==kNone) break;
  
  if (first_action)
  {
    first_action=false;
 
   if (!sleepat)
   {
    if (start_w_default)
    {
     int i=0;
     for (i=0;i<ARRAYLENGTH-1;i++)
     if (autoswitch_vals[i]==default_timespan) next_sleepmin_index=i-1;
    }
    else
    {
     next_sleepmin_index=0;//bit confusing; final value should be 1 but 1 will be added at the end of the method
    }
   }
   else //there is an active sleeptimer so starting from this we have to find the next higher value
   {
    next_sleepmin_index=0;
    int last_field=ARRAYLENGTH-1;
    if (autoswitch_vals[last_field]<0) last_field=ARRAYLENGTH-2;
 
    if (sleepat > autoswitch_vals[last_field]*60 + now) continue;
    
    int dest_index=-1;
    int i=0;
    for (;i<=last_field;i++)
    {
     if (sleepat >= autoswitch_vals[i]*60 + now) continue;
 
     dest_index=i;
     break;
    }
    if (dest_index==-1)
     continue;
    else
    {
     next_sleepmin_index=dest_index;
     continue;
    }
   }
  }  
  next_sleepmin_index+=1;

  if (next_sleepmin_index>=ARRAYLENGTH || autoswitch_vals[next_sleepmin_index]<0) next_sleepmin_index=0;
 }
 Interface->GetKey(false);
 lastaction=0;
}


eKeys cPluginSleeptimer::ShowMessage(cString msg)
{
 cSkinDisplayMessage *csdmsg=Skins.Current()->DisplayMessage();
 cSkinDisplay::Current()->SetMessage(mtInfo,*msg);
 cSkinDisplay::Current()->Flush();
   
 Interface->GetKey(false);
 eKeys retval= Interface->Wait(5);
     
 delete csdmsg;
       
 return retval;
}

VDRPLUGINCREATOR(cPluginSleeptimer); // Don't touch this!
